asynciterator.ts

```diff
@@ -1981,42 +1984,170 @@ class HistoryReader<T> {
   }
 }
 
+/**
+ * An iterator that takes a variety of iterable objects as a source.
+ */
+export class WrappingIterator<T> extends AsyncIterator<T> {
+  protected _source: InternalSource<T> | null = null;
+
+  constructor(source?: MaybePromise<IterableSource<T>>) {
+    super();
+    if (source) {
+      if (!isPromise(source))
+        this.source = source as any;
+      else
+        source.then(s => this.source = s).catch(e => this.emit('error', e));
+    }
+  }
+
+  set source(value: IterableSource<T>) {
+    let source: InternalSource<T> = value as any;
+    // Do not change sources if the iterator is already done
+    if (this.done)
+      return;
+
+    // Process an iterable source
+    if (isIterable(source))
+      source = source[Symbol.iterator]() as any;
+    // Process an iterator source
+    if (isIterator<T>(source)) {
+      let iterator: Iterator<T> | null = source;
+      source = new EventEmitter() as any;
+      source.read = (): T | null => {
+        if (iterator !== null) {
+          // Skip any null values inside of the iterator
+          let next: IteratorResult<T>;
+          while (!(next = iterator.next()).done) {
+            if (next.value !== null)
+              return next.value;
+          }
+          // No remaining values, so stop iterating
+          this.close();
+          iterator = null;
```

Comment:
I don't think this does anything as the iterator variable is never read after this point